Add validation and tests for order breaking name_function

nirizr opened the pull request:
This bug was recently introduced by myself to dataframe.to_hdf and preexisted with bag.to_textfiles.
Saving to multiple files (as possible with asterisks in those functions) did not preserve the
partitions order.

nirizr commented:
This PR is currently here only to demonstrate and highlight the issue, and fails on both of the new tests it implements.

I'll be fixing the issue soon.

mrocklin commented on dask/bag/core.py:
```diff
@@ -156,12 +158,15 @@ def to_textfiles(b, path, name_function=str, compression='infer',
+
+        formatted_names = [name_function(i) for i in range(b.npartitions)]
+        if formatted_names != sorted(formatted_names):
+            raise ValueError("name_function must preserve the order of its input")
```
I suggest that we warn instead of raise.  Generally I trust the user to make this decision on their own.

mrocklin commented on dask/bag/core.py:
```diff
@@ -156,12 +158,15 @@ def to_textfiles(b, path, name_function=str, compression='infer',
-            paths = [os.path.join(path, '%s.part' % name_function(i))
-                     for i in range(b.npartitions)]
+        if not '*' in path:
+            path = os.path.join(path, '*.part')
```
Nice

nirizr commented:
That's my suggestion for building `name_function` in `to_textfiles` and `to_hdf`.

The third commit tries to cleanup some of the tests i recently added, i can make it in a new PR.

The forth is removing some tests i would have wanted in, because python 2.7 uses an old version of pytest, that doesn't implement `pytest.warns`. I can work around it by catching the warn and asserting on it, but advice would be appreciated.

mrocklin commented on dask/bag/tests/test_bag.py:
```diff
@@ -647,6 +645,16 @@ def test_to_textfiles():
+    with tmpdir() as dn:
+        b.to_textfiles(dn)
+
+        # TODO: why do i need to map(str).map(str.strip)? how come there're unneeded newlines?
```
Perhaps remove the TODO and raise an issue?

mrocklin commented on dask/dataframe/tests/test_io.py:
```diff
@@ -697,25 +697,37 @@ def test_to_hdf():
+    with tmpfile('h5') as fn:
+        b.to_hdf(fn, '/data*')
+        out = dd.read_hdf(fn, '/data*')
+        tm.assert_frame_equal(df16, out.compute())
```
I recommend the `eq` function present in this module.  It will check some other handy things

```python
eq(df16, out)
```

mrocklin commented on dask/utils.py:
```diff
@@ -625,3 +626,33 @@ def insert(tup, loc, val):
+    '0'
+
+    """
+
```
There are a couple of extra lines there that should probably be removed

mrocklin commented:
This looks really good to me.  A couple of small nit-picky comments but honestly if you feel like ignoring these that'd be fine with me.

nirizr commented:
No, these are good comments. I'll resolve those.

I still have one issue i'd like your opinion, what do you think about testing warnings in python 3.3? (see commit 1c4d088)

mrocklin commented:
I would keep the tests but guard them with `if sys.version_info >= (3, 4):`

nirizr commented:
How's it looking now?
Fixed the three comments, replaced `tm.assert_frame_equal` with `eq` in many places, and I'll open an issue about the new lines in dask.bag later.

mrocklin commented on dask/bag/tests/test_bag.py:
```diff
@@ -647,6 +646,22 @@ def test_to_textfiles():
+    with tmpdir() as dn:
+        b.to_textfiles(dn)
+
+        out = db.read_text(dn+"/*", encoding='ascii').map(str).map(str.strip).compute()
```
Rather than `+ "/*"` I recommend `os.path.join(dn, '*')`.  Windows may fail here otherwise.

nirizr replied:
done

mrocklin commented:
+1 from me.  This looks great

mrocklin commented:
Thanks @nirizr !  This is in.
